Use DI factories to allow nice registration of admin integrations

wouterj opened the pull request:
* Fixed the tests and did some clean-ups
 * Refactored the DI extension to be extendable by factories. This allows more config validation, a better config dumper, dynamic config options per admin integration and to add integrations from inside other bundles

/cc @dbu @ElectricMaxxx @fbourigault I think we should be able to quickly move admins to this bundle after this PR

dbu commented on src/CmfSonataAdminIntegrationBundle.php:
```diff
@@ -12,8 +12,16 @@
+        parent::build($container);
 
+        $extension = $container->getExtension('cmf_sonata_admin_integration');
+        $extension->registerAdminFactory(new Factory\SeoAdminFactory());
```
should we do class_exists for the bundle classes to decide which parts we support?

wouterj commented:
Updated this PR to only register installed bundles.

ElectricMaxxx commented:
there are still some failing tests.

You do #9, #10 and #11 in here, right?

dbu commented on src/DependencyInjection/CmfSonataAdminIntegrationExtension.php:
```diff
@@ -16,19 +16,46 @@
+            $this->factories = $factories;
+        } else {
+            $bundles = [
+                CmfSeoBundle::class => new Factory\SeoAdminFactory(),
```
or should we just make the factories param not optional. the extension is created by the bundle class, so unless people customized something, nothing bad happens. if we already have a stable release, lets deprecate not passing factories and plan to remove the = null in version 2.

wouterj replied:
BC with what exactly? There has been no stable release of this bundle yet.

dbu commented on src/DependencyInjection/Factory/AdminFactoryInterface.php:
```diff
@@ -0,0 +1,50 @@
+interface AdminFactoryInterface
+{
+    /**
+     * The configuration key used to reference the admin.
```
should we be explicit that this is the bundle class FQN? we rely on that in the DI extension class

wouterj replied:
Same confusion here

fbourigault commented on src/DependencyInjection/CmfSonataAdminIntegrationExtension.php:
```diff
@@ -16,19 +16,46 @@
+            ];
+
+            foreach ($bundles as $bundleFqcn => $factory) {
+                if (class_exists($bundleFqcn)) {
```
Is it not better to rely on enabled bundles instead of installed ones (i.e. using `kernel.bundles` parameter)?

wouterj commented on src/DependencyInjection/CmfSonataAdminIntegrationExtension.php:
```diff
@@ -16,19 +16,46 @@
+    private $factories;
+
+    /**
+     * @param null|AdminFactoryInterface[] $factories A list of Admin factories
```
No, factories are just Names mapped to admin factories (e.g. `'seo' => new Factory\SeoAdminFactory()`).

Your confusing `$bundles` with `$this->factories`.

dbu replied:
yep, re-read it and i think i get it now. but then the phpdoc should be "Map of config key to factory for that section." or something similar, right? the key does have a meaning.

imho we should either not use keys or remote the getKey method from the factory interface - having both seems redundant to me.

dbu commented on src/DependencyInjection/CmfSonataAdminIntegrationExtension.php:
```diff
@@ -68,4 +89,22 @@ public function getXsdValidationBasePath()
         return __DIR__.'/../Resources/config/schema';
     }
+
+    private function defineFactories(ContainerBuilder $container)
```
as there is no stable release, i propose we drop this method and just set up factories from the bundle class

dbu replied:
or we can make the extension self-configuring, but then we should also create the factories if additional factories have been registered with the public method. or remove the public method as i would think that if people do additional things, they can just use the DI of their own bundle and don't need to interact with this bundle

dbu commented on src/DependencyInjection/Configuration.php:
```diff
@@ -2,46 +2,52 @@
+
+    private function addBundlesSection(ArrayNodeDefinition $root)
+    {
+        $bundles = $root->children()->arrayNode('bundles')->children();
```
is this for forward compatibility to namespace everything about admins for other bundles under a `bundles` key? not sure we need it, but does not cost much and is the defensive option so we can add things if we ever need to. in short: i agree with this structure ;-)

dbu commented on src/DependencyInjection/Factory/SeoAdminFactory.php:
```diff
@@ -0,0 +1,55 @@
+        $bundles = $container->getParameter('kernel.bundles');
+        if (!isset($bundles['BurgovKeyValueFormBundle'])) {
+            throw new InvalidConfigurationException(
+                'To use advanced menu options, you need the burgov/key-value-form-bundle in your project.'
```
this is the seo admin. why menu options? and we seem to hard-require burgov/key-value-form-bundle the way this is as there is no check on any configuration the user might or might not have activated

dbu replied:
hm, this seems to be moved code from the original class. @ElectricMaxxx can you help us here, any idea what this was about? does anything (or everything?) in the seo admin depend on the key-value-form?

dbu commented on src/Resources/config/schema/sonata-admin-1.0.xsd:
```diff
@@ -8,14 +8,14 @@
     <xsd:complexType name="config">
         <xsd:sequence>
-            <xsd:element name="bundle" type="bundle" minOccurs="0" />
+            <xsd:element name="bundles" type="bundles" minOccurs="0" maxOccurs="1" />
```
as there is no release yet, i suggest we remove the 1.0

wouterj commented:
Updated the PR to remove the constructor from the extension and to always add the CMF bundles to the config when they are enabled. Furthermore, I also removed the version from the XSD.

Ready to merge imo (build for 3.2 is expected to fail, as SonataAdmin doesn't support 3.2 yet).

wouterj commented:
I'm merging this, to move forward with other admin integrations. The remaining comments about the SeoBundle can be fixed in another PR.

dbu commented:
thanks wouter. agree with merging. created #15 about the issue i saw
which is not introduced by this PR
